### DISCUSSION TITLE
Improve experience with already activated frames

### DISCUSSION THREAD

**Discussion by jonathanmfung** [alphapapa/activities.el#68] [Ideas]
I think the user experience for a non-tabs-bar workflow could be improved. For reference, I use SwayWM on Wayland with `system-configuration-options` as `-C --with-pgtk --with-native-compilation --with-json`.

Expected behavior (activities-tabs-mode ON)
1. In tab 1 `activities-resume` for activity "foo". This renames the tab to "a: foo".
2. In tab 2 `activities-resume` for activity "foo". This correctly switches to tab 1.

Unexpected behavior (activities-tabs-mode OFF)
1. In frame 1 `activities-resume` for activity "foo". This renames the Wayland window to "a: foo".
2. In frame 2 `activities-resume` for activity "foo". This does nothing.

I believe `activities-switch` is interchangeable for `activities-resume` for both step 2's as they both use `activities--switch`.

--

My first suggestion was to use "window urgency", but I'm not aware if Emacs supports this, at least portably ([SE ref](https://emacs.stackexchange.com/questions/56036/portable-way-of-setting-the-windows-urgent-hint)).

Then I looked into how you handle already activated frames. I don't even know if `select-frame` ([activities--switch](https://github.com/alphapapa/activities.el/blob/master/activities.el#L504)) works on Wayland ([similar bug-gnu-emacs ref](https://mail.gnu.org/archive/html/bug-gnu-emacs/2023-02/msg00390.html)). I relegate frame management to my WM, so don't really use `select/raise-frame`.

So my suggestion is to simply add a logging message in `activities--switch` that lets the user knows the selected activity is already activated _somewhere_.

**Comment by jdtsmith** [comment]
Why does `raise-frame` not work?   For me (on mac) it switches mac tabs (which emacs sees as frames).

**Comment by jonathanmfung** [comment]
I believe `raise-frame` doesn't work because I use a tiling WM, so there isn't usually a notion of front/back.

However I just tried `select-frame-by-name` which, for me, doesn't actually switch focus, but at least sets window/workspace urgency. AFAIK Wayland doesn't expose a native way for programs to force focus, so compositors need to implement their own. (aforementioned bug-gnu-emacs ref, [Gnome's](https://superuser.com/a/1627429), [Sway's](https://github.com/swaywm/sway/blob/36f54679935d6c6706c7d4d5532e80a21d0e54e9/sway/sway.5.scd?plain=1#L622), [Wayland spec](https://wayland.app/protocols/xdg-activation-v1#xdg_activation_v1:request:activate)).

`select-frame-by-name`  uses `select-frame-set-input-focus`, so now I propose changing `activities--switch`'s `select-frame` to that.

**Comment by jdtsmith** [comment]
`select-frame-set-input-focus` also raises, so I think this won't work generally because `activities--switch` is also called from a timer, to save the activity.  So you'd get spontaneous raising/focus change. 

One idea is to add an optional `raise-and-focus` argument to `activities--switch`, and use `select-frame-set-input-focus` in lieu of `select-frame` if this argument is non-nil, setting it only via `activities-new`, `activities-resume`, and `activities-switch`; i.e. "opt-in" to raising instead of using the dynamic binding.

**Comment by alphapapa** [comment]
It sounds like this behavior needs to be configurable, so probably the way to handle it would be for the user to add a function to `activities-after-resume-functions` that does whatever is desired (showing a message, calling a different frame-related function, etc).

**Comment by alphapapa** [comment]
Please see the new option, `activities-after-switch-functions`.  You can use that to customize what happens after switching to an activity.

**Comment by jdtsmith** [comment]
Maybe now you can remove the `raise-frame` hack from `activities--switch`, since frame people can add that to the hook (there could well be people who do _not_ want to raise resumed frame).

**Comment by alphapapa** [comment]
> Maybe now you can remove the `raise-frame` hack from `activities--switch`, since frame people can add that to the hook (there could well be people who do _not_ want to raise resumed frame).

`¯\_(ツ)_/¯ `

Seriously, though, I feel like that one should stay, because what good is a selected frame if it's buried beneath other frames or apps.  But if some users actually complain about it and want it removed or optional, we'll see.
